### PR TITLE
Fix CPU bug, overhaul model runner, and update to lightning >=2.0

### PR DESCRIPTION
**Pull request by wfondrie**
This is a huge PR that:

- Hopefully fixes CPU-related bugs (#106). It works locally for me, even with MPS disabled.
- Removes the `no_gpu` parameter, in favor of providing `accelerator` and `devices` parameters to allow users to select custom devices readily without negating `CUDA_VISIBLE_DEVICES`. I think this is the best comprimise for #173.
- Completely refactors `model_runner.py` into a class, `ModelRunner`. I found it annoying to have to change arguments for models and such in multiple spots, so I think this change will make it much more maintainable.
- Makes a bunch of minor tweaks required for Pytorch-Lighting 2.0+. One of those is changing how metric logging and PSM recording is done. PSM writing is now done using the `on_predict_batch_end` rather than `on_predict_epoch_end` because the latter seems to no longer receive the predict results. The newer version of Lightning doesn't allow for dictionary metrics to be logged in the way we were doing before, so please pay attention to the changes in review.
- The integration test was overhauled to make it useful. It now trains a tiny model, evaluates it on a tiny file, and predicts the sequences for the same file. 
- Limits prediction to using 1 device. 

I'm tagging both @bittremieux and @melihyilmaz to review this one since this one is so big and I don't want to mess something up 🙈.

### CODE REVIEW

**Comment by wfondrie** [comment]
All tests passing on Linux, Windows, and  MacOS CPU runners! 🎉

File: casanovo/denovo/model_runner.py [hunk 1]
```diff
@@ -23,307 +25,350 @@
...
+        Returns
+        -------
+        SpectrumIndex or AnnotatedSpectrumIndex
+            The spectrum index for training, evaluation, or inference.
+        """
+        ext = (".mgf", ".h5", ".hdf5")
+        if not annotated:
+            ext += (".mzml", ".mzxml")
+
+        if msg and msg[-1] != " ":
```
Comment (bittremieux) [inline casanovo/denovo/model_runner.py:321]: **nitpick:** Maybe use `strip()` instead to make sure there is no trailing whitespace? And then add a space in each of the longer messages, rather than in this sub-message. Virtually identical effect and this is the simplest piece of code, but somehow it just feels nicer.

**Review by bittremieux — COMMENTED** [review]
Very nice changes, I just have a few minor comments.

File: casanovo/denovo/model_runner.py [hunk 2]
```diff
@@ -23,307 +25,350 @@
...
+        self.initialize_model(train=False)
+
+        test_index = self._get_index(peak_path, False, "")
+        self.writer.set_ms_run(test_index.ms_files)
+        self.initialize_data_module(test_index=test_index)
+        self.loaders.setup(stage="test", annotated=False)
+
+        self.trainer.predict(self.model, self.loaders.test_dataloader())
+
+    def initialize_trainer(self, train: bool) -> None:
```
Comment (bittremieux) [inline casanovo/denovo/model_runner.py:163]: **issue:** This `train` parameter is not used.

File: casanovo/casanovo.py [hunk 3]
```diff
@@ -159,18 +161,17 @@ def main(
...
-        logger.info("Evaluate a trained Casanovo model.")
-        model_runner.evaluate(peak_path, model, config)
-    elif mode == "train":
-        logger.info("Train the Casanovo model.")
-        model_runner.train(peak_path, peak_path_val, model, config)
+    with ModelRunner(config, model) as model_runner:
+        if mode == "denovo":
+            logger.info("Predict peptide sequences with Casanovo.")
+            model_runner.predict(peak_path, output)
+            model_runner.writer.save()
```
[the remark this answers is not in the document]
Comment (wfondrie) [reply casanovo/casanovo.py:168]: Yeah that seems like a good idea 👍

File: casanovo/denovo/model_runner.py [hunk 4]
```diff
@@ -23,307 +25,350 @@
...
-    else:
-        if not os.path.isfile(model_filename):
+        if train and any(from_scratch):
+            self.model = Spec2Pep(**model_params)
+            return
+        elif self.model_filename is None:
+            logger.error("A model file must be proided")
+            raise ValueError("A model file must be provided")
+
+        if not self.model_filename.exists():
```
Comment (bittremieux) [inline casanovo/denovo/model_runner.py:241]: **issue:** I think that `self.model_filename` is a string, so `exists()` is not a valid function.

File: casanovo/config.yaml [hunk 5]
```diff
@@ -119,5 +119,11 @@ model_save_folder_path: ""
 save_weights_only: True
 # Model validation and checkpointing frequency in training steps
 every_n_train_steps: 50_000
-# Disable usage of a GPU (including Apple MPS):
-no_gpu: False
+# The hardware accelerator to use. Must be one of:
+# “cpu”, “gpu”, “tpu”, “ipu”, “hpu”, “mps”, or “auto”
```
[the remark this answers is not in the document]
Comment (wfondrie) [reply casanovo/config.yaml:123]: Ahh yeah - those were pasted from the lightning docs 🤦‍♂️

[the remark this answers is not in the document]
Comment (wfondrie) [reply casanovo/config.yaml:123]: I may try removing the UTF-8 specification now. I didn't put two and two together as to why it suddenly wasn't working on Windows.

File: tests/test_integration.py [hunk 6]
```diff
@@ -1,26 +1,76 @@
+import functools
```
[the remark this answers is not in the document]
Comment (wfondrie) [reply tests/test_integration.py:1]: haha I normally have Ruff/isort setup to reorder everything for me, if you can't tell.

File: casanovo/denovo/model_runner.py (hunk 1 above)
Comment (wfondrie) [reply casanovo/denovo/model_runner.py:321]: Absolutely a better solution!

File: casanovo/denovo/model_runner.py (hunk 4 above)
Comment (wfondrie) [reply casanovo/denovo/model_runner.py:241]: Good catch! I've added a unit test covering this as well.

**Review by bittremieux — APPROVED** [review]

**Review by melihyilmaz — COMMENTED** [review]
I stumbled upon one issue when testing locally, which should be fixed with my recent commit,  but everything else looks great. @wfondrie Can you take a look at the only failing test  - I wasn't sure if we need to tweak test or my recent commit ? Feel free to merge afterwards!

Thanks!

File: casanovo/denovo/model_runner.py [hunk 7]
```diff
@@ -23,307 +25,350 @@
...
-    )
-    # Train the model.
-    trainer.fit(
-        model, train_loader.train_dataloader(), val_loader.val_dataloader()
-    )
-    # Clean up temporary files.
-    tmp_dir.cleanup()
+            raise FileNotFoundError("Could not find the model weights file")
+
+        self.model = Spec2Pep().load_from_checkpoint(
```
[the remark this answers is not in the document]
Comment (bittremieux) [reply casanovo/denovo/model_runner.py:263]: @melihyilmaz What does this FIXME mean?
> # FIXME: Handle the case for mps separately.
Can we just address it rather than leave this as an open todo?

[the remark this answers is not in the document]
Comment (bittremieux) [reply casanovo/denovo/model_runner.py:263]: Also, does the GitHub CI provide different devices to write a unit test for this issue, or is it CPU-only? [This](https://github.com/pytorch/pytorch/issues/61654) might be useful.

[the remark this answers is not in the document]
Comment (melihyilmaz) [reply casanovo/denovo/model_runner.py:263]: re: MPS case, I couldn't find how to obtain from the `Trainer` whether an mps accelerator is being used to pass that on to `map_location`

[the remark this answers is not in the document]
Comment (melihyilmaz) [reply casanovo/denovo/model_runner.py:263]: re: unit testing devices, I think the weight loading tests are disabled now but with weights matching the architecture, this case will automatically be tested as gpu-trained weights would be loaded into the model which is run on a cpu.

[the remark this answers is not in the document]
Comment (wfondrie) [reply casanovo/denovo/model_runner.py:263]: Why not just always load to CPU, then transfer to whatever device for training?

[the remark this answers is not in the document]
Comment (melihyilmaz) [reply casanovo/denovo/model_runner.py:263]: @wfondrie Meaning we set `map_location_device = "cpu"` by default and then move the model weights to whatever device in `trainer`? (if so some quick googling didn't yield how to do the latter)

[the remark this answers is not in the document]
Comment (bittremieux) [reply casanovo/denovo/model_runner.py:263]: If this leads to an error, I do think it's good to have an explicit unit test for it, rather than have it implied by the weights loading test. Good suggestion by @wfondrie.

**Comment by wfondrie** [comment]
Ok, so I think I have something that'll work pretty robustly: the model weights are loaded onto the current PyTorch default device, which is normally CPU. However, this let's us test it by changing the default device to the `meta` device that @bittremieux mentioned, which I've done here.

Also, I went ahead and included the initialization parameters with the model weights, so that loading weights is independent of the configuration provided --- the model will always match the loaded weights, except for the event of major architecture changes (Issue #156)

**Comment by wfondrie** [comment]
Good thing the integration tests were working, because it helped me catch a bug!

Anyway, I also changed checkpointing behavior to only keep the top 5 checkboints based on validation CE loss, but changed the `save_models` parameter to `save_top_k`, allowing us to set it at -1 if we reeeeally want all the checkpoints.

**Review by bittremieux — APPROVED** [review]
@wfondrie updates look good to me.

File: casanovo/denovo/model_runner.py [hunk 8]
```diff
@@ -54,11 +54,13 @@ def __init__(
         self.writer = None
 
         # Configure checkpoints.
-        if config.save_model:
+        if config.save_top_k is not None:
             self.callbacks = [
                 pl.callbacks.ModelCheckpoint(
                     dirpath=config.model_save_folder_path,
-                    save_top_k=-1,
+                    monitor="valid_CELoss",
```
Comment (melihyilmaz) [inline casanovo/denovo/model_runner.py:61]: Currently the model doesn't save the weights at the first checkpoint (determined by `every_n_train_steps`) despite running validation and only does so at subsequent checkpoints. This warning is the only trace of something going on under the hood:

> 2023-05-10 12:47:47,393 WARNING [py.warnings/MainProcess] warnings._showwarnmsg : /net/noble/vol1/home/melih/anaconda3/envs/casanovo_dev_v2/lib/python3.8/site-packages/lightning/pytorch/utilities/data.py:76: UserWarning: Trying to infer the `batch_size` from an ambiguous collection. The batch size we found is 128. To avoid any miscalculations, use `self.log(..., batch_size=batch_size)`.
  warning_cache.warn(

I think this has to do with the order of validation with its logging and checkpointing (i.e. checkpointing is attempted before the results from 1st validation are logged) but I'm not entirely sure how to fix it. In the worst case, current checkpointing but might be lagging behind validation by 1 step - leading to a mismatch between validation results and checkpoints.

**Review by melihyilmaz — COMMENTED** [review]
Thanks @wfondrie, I've tested locally and identified only two issues that arose with the recent commits.

File: casanovo/denovo/model_runner.py (hunk 8 above)
Comment (wfondrie) [reply casanovo/denovo/model_runner.py:61]: I think I was able to fix this - @melihyilmaz can you try it out and see if you observe the correct behavior?

File: casanovo/denovo/model_runner.py [hunk 9]
```diff
@@ -247,23 +246,9 @@ def initialize_model(self, train: bool) -> None:
...
-            map_location_device = "ipu"
-        # FIXME: Handle the case for mps separately.
-        else:
-            map_location_device = "cpu"
-
-        self.model = Spec2Pep().load_from_checkpoint(
+        self.model = Spec2Pep.load_from_checkpoint(
             self.model_filename,
-            map_location=torch.device(map_location_device),
-            **model_params,
```
[the remark this answers is not in the document]
Comment (wfondrie) [reply casanovo/denovo/model_runner.py]: This was intentional, but I've revised it to where it should handle both cases now. It'll first look for the model architecture in the checkpoint, then use the configuration file if they're missing.

**Review by melihyilmaz — APPROVED** [review]
No more issues, good to merge!
